sw/device/lib/dif/dif_padctrl.c

```diff
@@ -0,0 +1,180 @@
+
+  // Reset all DIO pad attributes to the default value
+  for (int i = 0; i < PADCTRL_DIO_PADS_MULTIREG_COUNT; ++i) {
+    // TODO: Fix reggen so DIO_PADS defines don't switch between having and not
+    // having a 0 depending on whether multiple regs are needed.
+    ptrdiff_t dio_pads_register_offset =
+        PADCTRL_DIO_PADS_REG_OFFSET + (i * PADCTRL_REGISTER_WIDTH_IN_BYTES);
+    mmio_region_write32(padctrl->base_addr, dio_pads_register_offset,
+                        PADCTRL_ATTR_DEFAULT);
+  }
+
+  // Reset all MIO pad attributes to the default value
+  for (int i = 0; i < PADCTRL_MIO_PADS_MULTIREG_COUNT; ++i) {
+    // TODO: Fix reggen so MIO_PADS defines don't switch between having and not
+    // having a 0 depending on whether multiple regs are needed.
+    ptrdiff_t mio_pads_register_offset =
+        PADCTRL_MIO_PADS0_REG_OFFSET + (i * PADCTRL_REGISTER_WIDTH_IN_BYTES);
+    mmio_region_write32(padctrl->base_addr, mio_pads_register_offset,
+                        PADCTRL_ATTR_DEFAULT);
+  }
+
+  return kDifPadctrlInitErrorNone;
+}
+
+dif_padctrl_registers_lock_error_t dif_padctrl_registers_lock(
+    dif_padctrl_t *padctrl) {
+  if (padctrl == NULL) {
+    return kDifPadctrlRegistersLockErrorInvalidPadctrl;
+  }
+
+  mmio_region_write32(padctrl->base_addr, PADCTRL_REGEN_REG_OFFSET, 0);
+
+  return kDifPadctrlRegistersLockErrorNone;
+}
+
+dif_padctrl_enable_attr_error_t dif_padctrl_enable_attr(
+    const dif_padctrl_t *padctrl, dif_padctrl_pad_kind_t kind,
+    dif_padctrl_pad_index_t index, dif_padctrl_attr_t attr, bool enable) {
+  if (padctrl == NULL) {
+    return kDifPadctrlEnableErrorInvalidPadctrl;
+  }
+
+  if (attr > kDifPadctrlAttrLast) {
+    return kDifPadctrlEnableErrorInvalidAttr;
+  }
+
+  ptrdiff_t pad_base_register_offset;
+  uint32_t pad_register_index;
+  uint32_t pad_attr_field_index;
+
+  // Do pad index range check and calculate which register and attribute field
+  // within it needs to be modified based on the kind of pad.
+  if (kind == kDifPadctrlPadMio) {
+    if (index > (PADCTRL_PARAM_NMIOPADS - 1)) {
+      return kDifPadctrlEnableErrorInvalidPad;
+    }
+
+    pad_base_register_offset = PADCTRL_MIO_PADS0_REG_OFFSET;
+    pad_register_index = index / PADCTRL_MIO_PADS_ATTR_FIELDS_PER_REG;
+    pad_attr_field_index = index % PADCTRL_MIO_PADS_ATTR_FIELDS_PER_REG;
+  } else {
+    if (index > (PADCTRL_PARAM_NDIOPADS - 1)) {
+      return kDifPadctrlEnableErrorInvalidPad;
+    }
+
+    pad_base_register_offset = PADCTRL_DIO_PADS_REG_OFFSET;
+    pad_register_index = index / PADCTRL_DIO_PADS_ATTR_FIELDS_PER_REG;
+    pad_attr_field_index = index % PADCTRL_DIO_PADS_ATTR_FIELDS_PER_REG;
+  }
+
+  // Determine which bit the attribute field that is being modified starts at
+  // and which bit within the register controls the attribute being modified.
+  uint32_t attr_field_start_bit = pad_attr_field_index * PADCTRL_PARAM_ATTRDW;
+  uint32_t attr_bit_index = attr_field_start_bit + attr;
```

Comment:
I think you don't need this, you would be able to directly use `mmio_region_nonatomic_set_field32` or `bitfield_set_field32`.

Reply:
The issue here is the need to read the register, check the bits in the attribute field for conflicts, then write as explained in the other comment. However using the bitfield_X functions should work and would be better, I will switch to using them.

Reply:
Switched to the bitfield functions so most of the raw bit manip stuff is gone.